net/net processing: Move tx inventory into net_processing

jnewbery opened the pull request:
This continues the work of moving application layer data into net_processing, by moving all tx data into the new Peer object added in #19607.

For motivation, see #19398.

jnewbery committed:
scripted-diff: rename TxRelay members

-BEGIN VERIFY SCRIPT-
ren() { sed -i "s:\<$1\>:$2:g" $(git grep -l "\<$1\>" ./src ./test); }

ren cs_filter             m_bloom_filter_mutex
ren fRelayTxes            m_relay_txs
ren pfilter               m_bloom_filter
ren cs_tx_inventory       m_tx_inventory_mutex
ren filterInventoryKnown  m_tx_inventory_known_filter
ren setInventoryTxToSend  m_tx_inventory_to_send
ren fSendMempool          m_send_mempool
ren nNextInvSend          m_next_inv_send_time
ren minFeeFilter          m_fee_filter_received
ren lastSentFeeFilter     m_fee_filter_sent
-END VERIFY SCRIPT-

jnewbery committed:
[net processing] Move tx relay data to Peer

jnewbery committed:
[net] Add CNode.m_relays_txs and CNode.m_bloom_filter_loaded

We'll move the transaction relay data into Peer in subsequent commits,
but the inbound eviction logic needs to know if the peer is relaying
txs and if the peer has loaded a bloom filter.

This is currently redundant information with m_tx_relay->fRelayTxes,
but when m_tx_relay is moved into net_processing, then we'll need these
separate fields in CNode.

jnewbery committed:
[net processing] Move m_wtxid_relay to Peer

Also, remove cs_main guard from m_wtxid_relay_peers and make it atomic.
This should be fine since we don't need m_wtxid_relay_peers to be
synchronized with m_wtxid_relay exactly at all times.

After this change, RelayTransaction no longer requires cs_main.

jnewbery commented:
Rebased on master to pick up fix for interface_zmq.py in #21008.

jnewbery commented:
Rebase resulted in silent merge conflict with 9476886353dffb730dcb75799f2bd5e143425795. I'll fix tomorrow.

jnewbery commented:
I've rebased this and moved it out of draft. It's now ready for review.

vasild reviewed (commented):
It seems sub-optimal that new fields are added - `CNode::m_relays_txs` and `CNode::m_bloom_filter_loaded` that are redundant/duplicates of `Peer::TxRelay::m_relay_txs` and `Peer::TxRelay::m_bloom_filter != nullptr`. This could be confusing and leaves a chance that in the future they go out of sync. Can the duplicates be avoided somehow?

There are two user-visible changes that may be considered as regressions in the [UI](https://github.com/bitcoin/bitcoin/pull/21160#discussion_r597831033) and in [RPC](https://github.com/bitcoin/bitcoin/pull/21160#discussion_r597832782).

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1460,17 +1508,16 @@ void PeerManagerImpl::SendPings()
-        } else {
-            pnode->PushTxInventory(txid);
+    LOCK(m_peer_mutex);
+    for(auto& it : m_peer_map) {
+        Peer& peer = *it.second;
```
That means node_id is unused, which causes warnings depending on your compiler flags.

vasild replied:
Which compiler flags produce a warning? I don't get any with either clang 12 or gcc 10:
```cpp
    std::map<int, int> m;
    for (const auto& [x, y] : m) { // no warning
        std::cout << x;
    } 
    for (const auto& [x, y] : m) { // warning: unused structured binding declaration [-Wunused-variable]
    } 
```

jnewbery replied:
Hmmm, if you search for "c++ structured binding unused variable", there do seem to be a lot of people claiming that they've encountered this warning. Like you, I've been unable to reproduce it, either locally or on godbold.org with various compiler versions.

In any case, I'm going to leave this as it is, since it's only one more line and no less clear to explicitly declare the `Peer` object. Thank you for pointing this out though!

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1460,17 +1508,16 @@ void PeerManagerImpl::SendPings()
+    LOCK(m_peer_mutex);
+    for(auto& it : m_peer_map) {
+        Peer& peer = *it.second;
+        if (!peer.m_tx_relay) return;
```
We don't have a project style on whether to compare to `nullptr` or use a pointer's bool conversion. Both are fine.

jnewbery commented:
Rebased

MarcoFalke commented:
Doesn't compile

jnewbery commented:
Thanks Marco. Bad rebase should be fixed now.

jnewbery commented:
Thanks for the review @vasild! I think I've addressed all of your review comments now.

jnewbery commented on src/rpc/net.cpp:
```diff
@@ -193,7 +193,6 @@ static RPCHelpMan getpeerinfo()
         }
         obj.pushKV("services", strprintf("%016x", stats.nServices));
         obj.pushKV("servicesnames", GetServicesNames(stats.nServices));
-        obj.pushKV("relaytxes", stats.fRelayTxes);
```
As above for the gui peer console.

jnewbery commented on src/qt/rpcconsole.cpp:
```diff
@@ -1114,7 +1114,6 @@ void RPCConsole::updateDetailWidget()
         peerAddrDetails += "<br />" + tr("via %1").arg(QString::fromStdString(stats->nodeStats.addrLocal));
     ui->peerHeading->setText(peerAddrDetails);
     ui->peerServices->setText(GUIUtil::formatServicesStr(stats->nodeStats.nServices));
-    ui->peerRelayTxes->setText(stats->nodeStats.fRelayTxes ? ts.yes : ts.no);
```
In reality, I think `fNodeStateStatsAvailable` will almost always be true, so this isn't a big issue (if it was, then the other stats in `nodeStateStats` not being available would also be an issue).

jonatack replied:
Oh, this is a bit surprising.  FWIW I found that I needed to add null-checking for the new `m_addr_*` statestats in #22501 to avoid parsing errors and may propose a fix in getpeerinfo.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -4443,83 +4492,84 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
                     } else {
-                        pto->m_tx_relay->nNextInvSend = PoissonNextSend(current_time, OUTBOUND_INVENTORY_BROADCAST_INTERVAL);
+                        // Use half the delay for outbound peers, as there is less privacy concern for them.
+                        peer->m_tx_relay->m_next_inv_send_time = PoissonNextSend(current_time, OUTBOUND_INVENTORY_BROADCAST_INTERVAL);
```
Oops! Thank you. Removed.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -610,6 +643,14 @@ static CNodeState *State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
     return &it->second;
 }
 
+static void AddKnownTx(Peer& peer, const uint256& hash)
+{
+    if (peer.m_tx_relay != nullptr) {
+        LOCK(peer.m_tx_relay->m_tx_inventory_mutex);
+        peer.m_tx_relay->m_tx_inventory_known_filter.insert(hash);
+    }
+}
```
I'm trying to keep `Peer` as a data structure without any internal logic.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -217,7 +250,10 @@ struct Peer {
     std::deque<CInv> m_getdata_requests GUARDED_BY(m_getdata_requests_mutex);
 
-    explicit Peer(NodeId id) : m_id(id) {}
+    explicit Peer(NodeId id, bool blocks_only)
```
Nice! Removed.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2597,9 +2645,8 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             return;
         }
         if (pfrom.GetCommonVersion() >= WTXID_RELAY_VERSION) {
-            LOCK(cs_main);
-            if (!State(pfrom.GetId())->m_wtxid_relay) {
-                State(pfrom.GetId())->m_wtxid_relay = true;
+            if (!peer->m_wtxid_relay) {
+                peer->m_wtxid_relay = true;
                 m_wtxid_relay_peers++;
```
Yes, that's a good observation. I agree that this is totally fine, and the fields don't need to be atomically updated under cs_main.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1006,7 +1047,6 @@ void PeerManagerImpl::FinalizeNode(const CNode& node, bool& fUpdateConnectionTim
     m_outbound_peers_with_protect_from_disconnect -= state->m_chain_sync.m_protect;
     assert(m_outbound_peers_with_protect_from_disconnect >= 0);
-    m_wtxid_relay_peers -= state->m_wtxid_relay;
     assert(m_wtxid_relay_peers >= 0);
```
Done.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -983,6 +1023,7 @@ void PeerManagerImpl::FinalizeNode(const CNode& node, bool& fUpdateConnectionTim
         PeerRef peer = RemovePeer(nodeid);
         assert(peer != nullptr);
         misbehavior = WITH_LOCK(peer->m_misbehavior_mutex, return peer->m_misbehavior_score);
+        m_wtxid_relay_peers -= peer->m_wtxid_relay;
```
Done.

vasild replied:
Now the move of the decrement happens in commit
`[net processing] Move m_wtxid_relay to Peer`
while the move of the assert happens in
`[net processing] Move tx relay data to Peer`.

I think both moves belong to the first commit.

jnewbery replied:
Oops. Now fixed. Thank you!

jnewbery commented on src/net.cpp:
```diff
@@ -974,18 +963,11 @@ bool CConnman::AttemptToEvictConnection()
-            }
             NodeEvictionCandidate candidate = {node->GetId(), node->nTimeConnected, node->m_min_ping_time,
                                                node->nLastBlockTime, node->nLastTXTime,
                                                HasAllDesirableServiceFlags(node->nServices),
-                                               peer_relay_txes, peer_filter_not_null, node->nKeyedNetGroup,
-                                               node->m_prefer_evict, node->addr.IsLocal()};
+                                               node->m_relays_txs.load(), node->m_bloom_filter_loaded.load(),
+                                               node->nKeyedNetGroup, node->m_prefer_evict, node->addr.IsLocal()};
             vEvictionCandidates.push_back(candidate);
```
Seems reasonable, but I agree it's out of scope for this PR.

jnewbery commented on src/net.cpp:
```diff
@@ -565,12 +565,6 @@ void CNode::copyStats(CNodeStats &stats, const std::vector<bool> &m_asmap)
     X(addrBind);
```
Thanks! Fixed.

jnewbery commented on src/net.h:
```diff
@@ -551,34 +549,14 @@ class CNode
-        CAmount lastSentFeeFilter{0};
-        std::chrono::microseconds m_next_send_feefilter{0};
-    };
+    /** Whether this peer asked us to relay transactions. This only changes from
```
That's better. I've updated the comment.

jnewbery commented:
> It seems sub-optimal that new fields are added - CNode::m_relays_txs and CNode::m_bloom_filter_loaded that are redundant/duplicates of Peer::TxRelay::m_relay_txs and Peer::TxRelay::m_bloom_filter != nullptr. This could be confusing and leaves a chance that in the future they go out of sync. Can the duplicates be avoided somehow?

That does seem a bit of a shame, but note that even though the data seems redundant, it's used for different purposes in net_processing and net. TxRelay in net_processing is data that is used in relaying transactions to that peer. `m_relay_txs` and `m_bloom_filter` in net are simply metrics used in the eviction logic. net_processing informs net of any events that might be used in net's eviction logic (e.g. min ping time, last tx received, last block received), and these new metrics (m_relay_txs and m_bloom_filter) are no different from those.

vasild commented:
Reviewed 66baf28fe2b76d3f214714604e9e24e83f16c20e - the patch looks correct and I do not think it introduces bugs.

I will try to see when the [UI](https://github.com/bitcoin/bitcoin/pull/21160#discussion_r597831033) and [RPC](https://github.com/bitcoin/bitcoin/pull/21160#discussion_r597832782) fields could become empty/missing when they would have been filled/present before this PR.

jnewbery commented:
Thanks for the re-review @vasild. I've updated the commits to fix https://github.com/bitcoin/bitcoin/pull/21160#discussion_r599387582.

jnewbery commented:
Rebased

jnewbery commented:
rebased

jnewbery commented:
Rebased

glozow commented on src/net_processing.cpp:
```diff
@@ -218,7 +251,10 @@ struct Peer {
     std::deque<CInv> m_getdata_requests GUARDED_BY(m_getdata_requests_mutex);
 
-    explicit Peer(NodeId id) : m_id(id) {}
+    Peer(NodeId id, bool blocks_only)
```
Why'd you delete the `explicit`?

jnewbery replied:
No good reason. I've put it back.

vasild replied:
`explicit` prevents implicit conversions for _single-argument_ constructors. After the change above, the constructor has 2 arguments. See https://stackoverflow.com/a/121163

MarcoFalke replied:
`explicit` will also prevent implicit conversion for multiple-argument constructors. Is there a reason why {int,bool} should implicitly be converted into Peer?

MarcoFalke replied:
Was this addressed?

jnewbery replied:
fixed now

glozow commented on src/net_processing.cpp:
```diff
@@ -1470,17 +1518,16 @@ void PeerManagerImpl::SendPings()
+    LOCK(m_peer_mutex);
+    for(auto& it : m_peer_map) {
+        Peer& peer = *it.second;
+        if (!peer.m_tx_relay) return;
```
Why is this `return` instead of `break`? Maybe I'm misunderstanding what this is doing, but I thought that we'd continue iterating through the others to relay to them, but returning seems like we're just skipping over the rest?

jnewbery replied:
Yikes. Good catch. In a previous iteration of this PR, I had a `ForEachPeer()` helper that took a lambda, and this `return` was effectively a `continue` (the lambda would return and then be called for the remaining peers). Now fixed!

jnewbery commented on src/net_processing.cpp:
```diff
@@ -209,6 +209,39 @@ struct Peer {
     std::atomic<bool> m_ping_queued{false};
 
+    /** Whether this peer relays txs via wtxid */
+    bool m_wtxid_relay{false};
```
Correct. The `Peer` object is constructed before we receive the `wtxidrelay` message, so we can't make this const.

jnewbery commented:
Thanks for the review @glozow. I've rebased on master and addressed your review comments.

#21527 moves around the locks in net_processing, which interacts a bit with how this PR moves `m_wtxid_relay` around. I think it makes sense to focus on merging that PR first, since that could simplify the locking here a bit.

jnewbery commented:
I'll rebase this after #21527 is merged

jnewbery commented:
I've rebased this on top of the _net_processing: add m_mutex_message_handling_ commit from #21527. I'm leaving as WIP for now, but it's in a state where people can review it. I'll move it out of draft as soon as 21527 or #22053 is merged (should be a trivial rebase).

jnewbery commented:
I'd rebased this on top of some of the work in #21527, since they conflicted a little. However, it seems that 21527 has now been abandoned, so I've removed that commit and marked this PR ready for review again. If 21527 does get picked up again and merged, then it should be easy enough to rebase this on top of it.

jnewbery commented:
Rebased on master

glozow commented on src/net_processing.cpp:
```diff
@@ -227,28 +227,28 @@ struct Peer {
         //    unless it loads a bloom filter.
-        bool fRelayTxes GUARDED_BY(cs_filter){false};
-        std::unique_ptr<CBloomFilter> pfilter PT_GUARDED_BY(cs_filter) GUARDED_BY(cs_filter){nullptr};
+        bool m_relay_txs GUARDED_BY(m_bloom_filter_mutex){false};
```
In 2533ea5161d2cb15b3a43e0fea9b146f525aea97

Noting that we have 3 variables named `m_relay_txs`: in `NodeEvictionCandidate`, `Peer.TxRelay`, and `CNodeStateStats`. And we have a `CNode.m_relays_txs` (very similar but not same name) where `CNode.m_relays_txs == Peer.m_tx_relay.m_relay_txs` always, and they are not the same thing as `Peer.m_tx_relay != nullptr`.

jnewbery replied:
Yes, this is slightly unfortunate. Please let me know if you have suggestions for improving the naming.

MarcoFalke replied:
Is there a reason for the bool to exist? What would happen if we delayed initialization of the structure to when the version message is received?

Something similar is done for address relay, IIRC.

jnewbery replied:
@MarcoFalke yes, that's the plan. This PR is intended to be a simple refactor, and delaying initialization of the `TxRelay` struct requires more focused discussion so should be done in a separate PR.

I've opened PR #22778, which implements delaying constructing `m_tx_relay` until the `version` message is received. Note however that `m_relay_txs` is still required - a peer that we've offered `NODE_BLOOM` services to may send us a version with `fRelay=0`, but later request that we start announcing transactions by sending `filterload` or `filterclear`, so we need to initialize the `TxRelay` struct during `version` handling but use `m_relay_txs` to gate whether we announce transactions.

I'm going to mark this conversation as resolved, but please let me know in #22778 if you have any other thoughts/feedback.

glozow reviewed (commented):
code review ACK 2533ea5161d2cb15b3a43e0fea9b146f525aea97

jnewbery commented on src/net_processing.cpp:
```diff
@@ -4710,83 +4745,83 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
             peer->m_blocks_for_inv_relay.clear();
         }
 
-        if (pto->m_tx_relay != nullptr) {
-                LOCK(pto->m_tx_relay->cs_tx_inventory);
+        if (peer->m_tx_relay != nullptr) {
+                LOCK(peer->m_tx_relay->m_tx_inventory_mutex);
                 // Check whether periodic sends should happen
                 bool fSendTrickle = pto->HasPermission(NetPermissionFlags::NoBan);
-                if (pto->m_tx_relay->nNextInvSend < current_time) {
+                if (peer->m_tx_relay->m_next_inv_send_time < current_time) {
                     fSendTrickle = true;
                     if (pto->IsInboundConn()) {
-                        pto->m_tx_relay->nNextInvSend = m_connman.PoissonNextSendInbound(current_time, INBOUND_INVENTORY_BROADCAST_INTERVAL);
+                        peer->m_tx_relay->m_next_inv_send_time = m_connman.PoissonNextSendInbound(current_time, INBOUND_INVENTORY_BROADCAST_INTERVAL);
                     } else {
-                        pto->m_tx_relay->nNextInvSend = PoissonNextSend(current_time, OUTBOUND_INVENTORY_BROADCAST_INTERVAL);
+                        peer->m_tx_relay->m_next_inv_send_time = PoissonNextSend(current_time, OUTBOUND_INVENTORY_BROADCAST_INTERVAL);
                     }
                 }
 
                 // Time to send but the peer has requested we not relay transactions.
                 if (fSendTrickle) {
-                    LOCK(pto->m_tx_relay->cs_filter);
-                    if (!pto->m_tx_relay->fRelayTxes) pto->m_tx_relay->setInventoryTxToSend.clear();
+                    LOCK(peer->m_tx_relay->m_bloom_filter_mutex);
+                    if (!peer->m_tx_relay->m_relay_txs) peer->m_tx_relay->m_tx_inventory_to_send.clear();
```
Hmmm, actually `m_tx_inventory_mutex` is held at this point (see line 4747 above):

```c++
        if (peer->m_tx_relay != nullptr) {
                LOCK(peer->m_tx_relay->m_tx_inventory_mutex);
                ...
```

In fact, all access of `m_tx_inventory_to_send` (and `setInventoryTxToSend` in master) is guarded by `m_tx_inventory_mutex` (`cs_tx_inventory` in master). It looks like an oversight in #13123 that a lock annotation wasn't added to `setInventoryTxToSend`. It could also be added to ` m_next_inv_send_time` (`nNextInvSend` in master).

We could add those annotations in a follow-up, since this branch is simply moving the data members and not changing any behavior here. I also have a separate follow-up branch that rationalizes these mutexes to a single mutex.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -3212,6 +3205,8 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         }
 
+        LOCK2(cs_main, g_cs_orphans);
+
```
Right. This `LOCK(cs_main)` was required to access the `CNodeState` in the line below. Now that `m_wtxid_relay` is in `Peer`, we don't need to hold `cs_main` until a little bit later (when we access `m_txrequest`).

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1199,6 +1195,8 @@ void PeerManagerImpl::FinalizeNode(const CNode& node)
         PeerRef peer = RemovePeer(nodeid);
         assert(peer != nullptr);
         misbehavior = WITH_LOCK(peer->m_misbehavior_mutex, return peer->m_misbehavior_score);
+        m_wtxid_relay_peers -= peer->m_wtxid_relay;
```
I agree that would be clearer.

vasild reviewed (commented):
2533ea5161d2cb15b3a43e0fea9b146f525aea97 looks fine on the low level.

It is best if internal refactors like this one do not introduce user visible changes (even less if those changes look like "regressions" -- missing fields whereas fields were present before) or code smells like the same data being stored in two different places (that can go out of sync). See an earlier [comment](https://github.com/bitcoin/bitcoin/pull/21160#pullrequestreview-616342920).

While those may look minor, it is difficult for me to judge whether the benefits of this PR outweight the above.

Wrt the missing fields in `getpeerinfo`: I think that may happen (`fStateStats` be `false`) at least in the case where the node is removed after `CConnMan::GetNodeStats()` is called, maybe in other cases too.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1666,22 +1706,17 @@ void PeerManagerImpl::SendPings()
-    WITH_LOCK(cs_main, _RelayTransaction(txid, wtxid););
-}
-
-void PeerManagerImpl::_RelayTransaction(const uint256& txid, const uint256& wtxid)
```
fixed.

jnewbery commented:
Thanks for the review @vasild. I understand your arguments about the extra fields in `CNode` and removing the stats from the GUI and RPC when they're not available. I don't have anything to add beyond my responses from the first time you raised those points.

jonatack commented:
> Thanks for the review @vasild. I understand your arguments about the extra fields in `CNode` and removing the stats from the GUI and RPC when they're not available.

(If helpful, mentioning https://github.com/bitcoin/bitcoin/pull/21160#discussion_r692284808 here, as that thread is closed as resolved.)

maflcko reviewed (commented):
Concept ACK. This will simplify `-nomempool`

glozow commented:
re ACK a192c23b2cb45e7c9a06bb00714347ef38df701f via `git range-diff 2533ea5...a192c23`, no overall changes, diff moved between 2 commits.

glozow commented:
reACK 34bd2711e7ffa0d5a7b563d15d79f8ae1d8cef43 via `git range-diff a192c23...34bd271`

hebasto commented:
Concept ACK.

MarcoFalke commented on src/test/fuzz/util.cpp:
```diff
@@ -204,18 +204,13 @@ void FillNode(FuzzedDataProvider& fuzzed_data_provider, CNode& node, bool init_v
     }
-    if (node.m_tx_relay != nullptr) {
-        LOCK(node.m_tx_relay->cs_filter);
-        node.m_tx_relay->fRelayTxes = filter_txs;
```
eeb0e84fc5c259dfd00f7673ad4a51582b100282: Would be nice to not break the fuzz test. Maybe a test-only mock method on PeerMan can do this?

jnewbery replied:
Does this break the fuzz test? `fRelayTxes` (`m_relay_txs` after this PR) will get set after the node receives and processes the `version` message. Presumably the fuzzer will be able to set the `fRelay` field on that message to true or false.

MarcoFalke replied:
Thanks, fixed in https://github.com/bitcoin/bitcoin/pull/23575

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -284,8 +314,9 @@ struct Peer {
     std::deque<CInv> m_getdata_requests GUARDED_BY(m_getdata_requests_mutex);
 
-    explicit Peer(NodeId id)
+    Peer(NodeId id, bool blocks_only)
```
eeb0e84fc5c259dfd00f7673ad4a51582b100282: Would be nice to rename this from `blocks_only` to `initialize_tx_relay` (or similar). `blocks_only` is confusingly similar named to the setting `-blocksonly`, but it really means block-relay-only.

Moreover, the details *why* tx relay isn't initialized here don't matter *here*, so a bool with a more general name seems fine.

jnewbery replied:
Done

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -1286,6 +1325,11 @@ bool PeerManagerImpl::GetNodeStateStats(NodeId nodeid, CNodeStateStats& stats) c
+    if (peer->m_tx_relay != nullptr) {
+        stats.fRelayTxes = WITH_LOCK(peer->m_tx_relay->cs_filter, return peer->m_tx_relay->fRelayTxes);
+        stats.minFeeFilter = peer->m_tx_relay->minFeeFilter.load();
+    }
```
eeb0e84fc5c259dfd00f7673ad4a51582b100282: any reason to make this code more fragile by removing the `else` in the move? The `else` used to set those to `false`, and `0`, respectively.

jnewbery replied:
I don't think I agree that this is more fragile. The default is that there connection isn't being used to relay transactions, and the feefilter is set to zero. Those stats are only updated if the node is relaying transactions/has set a feefilter.

I've reverted this to use if/else and remove the default initialization, since that's closer to the current code.

MarcoFalke commented on src/net_processing.h:
```diff
@@ -29,6 +29,8 @@ struct CNodeStateStats {
     std::chrono::microseconds m_ping_wait;
     std::vector<int> vHeightInFlight;
+    bool fRelayTxes{false};
+    CAmount minFeeFilter{0};
```
eeb0e84fc5c259dfd00f7673ad4a51582b100282: Why not `-1` (an impossible number), like the height values above? Or  maybe leave it uninitialized to make memory sanitizers more useful?

jnewbery replied:
As above (https://github.com/bitcoin/bitcoin/pull/21160#discussion_r715508638)

maflcko reviewed (commented):
Concept ACK 34bd2711e7ffa0d5a7b563d15d79f8ae1d8cef43 🤺

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Concept ACK 34bd2711e7ffa0d5a7b563d15d79f8ae1d8cef43 🤺
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUi+Jwv/Ur+qsT+NGY4mzH2EdkIAAFERJYtimCUi2QIug57wZnFMdbuGmmQ1rLH8
7VdAD3sG1SLDH76CkToopAYidMgxwpX2I07fV9YY+Rf3V3brS+aSSBJP0jLH0c2e
lHfrOA+kwK6L9tEBuODUBlAM3oKglqa8R50SUha9EHZ+dQGnkNufNtH7PA9XEBuS
HSZU4w2+u6j5IHvN94lgIe+HdympyeMK+JRvzOYZrX5IThzhZ3k1Ho5aoJtjv60b
FIPZ/oskBKG1dlnQU02OPZLwyjXOBAvxfiXpNHGEiBrUFQEmgLbldPXJsBGZtQv/
XN5S7YvSSpKSv7dMeGKUxbjesA5WYJH5/f5vzb2LADPhpdNIqbHn6aL/60mEKqpq
2++CjN5J7o9mUoGgVOlH2rHe/x8gu8kaSBRhLnDI7kkIjbxCfLJQ2nR3hjxo5URh
N602XW1Uxmos1hl0JsMRZ6GTzT5Zw9lPyA4q/JDqcirR0+RnUTBZStvAeAQhF4+B
lgygIrVL
=+2rA
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `98a6c2ba1a3b3b76dacfb7357c8a39198b4cbb57741f9a173f774636f7632c66  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010898a6c2ba1a3b3b76dacfb7357c8a39198b4cbb57741f9a173f774636f7632c66f010649183e3284956e0a365098a6e15b54d08fff01015927f6f5856bdcbe42d11a4a4dcc32f08f1046130dfb7f008f1ec83fb11cdb6c50083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010376cd6116cd82259a6550957d9c7cbd508f1046130dfb7f0080f883d7f35b529bb0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010c55466d63b78e46ac7e11f73283e87ab08f1046130dfb7f008e4633480d9dcce430083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010d62d01d1154d66cfa3dc0d2ecd958e8408f120935f1fc56526a77aa01a5ba5428de4a6ebd5721be339d31aa7e051dcb353d6a708f1046130dfb6f00833d7319c5212f3420083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented:
Are you still working on this?

jnewbery commented:
> Are you still working on this?

@MarcoFalke 

Yes. Apologies for the delay. I do intend to rebase and address your review comments soon.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2593,6 +2593,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         if (pfrom.m_tx_relay != nullptr) {
             LOCK(pfrom.m_tx_relay->cs_filter);
             pfrom.m_tx_relay->fRelayTxes = fRelay; // set to true after we get the first filter* message
+            if (fRelay) pfrom.m_relays_txs = true;
```
Because of the comment for m_relay_txs in net.h:

    /** ... This only changes from false to true. It will never change
     *  back to false. Used only in inbound eviction logic. */

This doesn't work:

```suggestion
            pfrom.m_relays_txs &= fRelay;
```

since the `&=` operator isn't defined for `std::atomic<bool>`. I could use:

```suggestion
            pfrom.m_relays_txs = pfrom.m_relay_txs && fRelay;
```

but I don't think that's an improvement.

vasild replied:
`pfrom.m_relays_txs = pfrom.m_relay_txs && fRelay;` could change it from `true` to `false`.

naumenkogs replied:
I think the comment is a good justification to keeping the code this way.
The only problem I can see is the temptation of future devs to change this.

jnewbery replied:
> pfrom.m_relays_txs = pfrom.m_relay_txs && fRelay; could change it from true to false.

:man_facepalming: You're right of course. I meant `pfrom.m_relay_txs || fRelay;`

jnewbery replied:
> Can you move this out of the lock scope to make that clear?

Done in #24692

jnewbery commented on src/net_processing.cpp:
```diff
@@ -223,6 +223,9 @@ struct Peer {
     std::atomic<bool> m_ping_queued{false};
 
+    /** Whether this peer relays txs via wtxid */
+    bool m_wtxid_relay{false};
```
No reason. I've updated this to be an atomic.

jnewbery commented:
Thanks for the review @MarcoFalke. I've addressed your review comments and rebased on master.

naumenkogs commented on src/net.h:
```diff
@@ -557,6 +557,16 @@ class CNode
 
+    /** Whether we should relay transactions to this peer (their version
+     *  message did not include fRelay=false and this is not a block-relay-only
+     *  connection). This only changes from false to true. It will never change
```
>This only changes from false to true

Could you mention that this happens only at the beginning of the connection lifetime?

jnewbery replied:
`m_relays_txs` can be set to true after initial connection, if the peer sends a `filterload` or `filterclear` message.

naumenkogs replied:
Right, thanks.

naumenkogs commented:
ACK ce772d53048a8d0d9934a8310dbedcbdafdc34ec
A good change for better readability with no behavior changes.

theuni reviewed (commented):
Concept ACK.

My only reservation here is that we lose a few sync guarantees now that some things are done outside of `cs_main`. For example `m_wtxid_relay_peers` and the _actual_ count may  momentarily disagree. Are you confident that nothing is relying on the assumption that they're in sync?

jnewbery commented:
rebased. I still intend to address @theuni's comments.

MarcoFalke commented:
This has been needing rebase and stuff addressed for more than a week, so I've removed it from hi-prio for now. (Can be added back later)

jnewbery commented:
> This has been needing rebase and stuff addressed for more than a week, so I've removed it from hi-prio for now. (Can be added back later)

Thanks @MarcoFalke. That's fair. I do intend to rebase this. Just have a lot going on right now.

MarcoFalke commented:
Absolutely. Take your time and maybe see you rebase this next year :heart:

jnewbery commented:
This is needed for some other work, so I've rebased on latest master. I still haven't addressed @theuni's review comments.

MarcoFalke commented:
re-ACK  0d025d56c1882ef1d4f3e0d78d337a158791f4d9 🚦

only changes are:

* Rebase
* Fix Peer constructor (tx_relay bool)


<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK  0d025d56c1882ef1d4f3e0d78d337a158791f4d9 🚦

only changes are:

* Rebase
* Fix Peer constructor (tx_relay bool)

-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjPdwv/ZLe2SAvo1xnZxY38S+0T2XwINcmvufQZrieVafJfRGQaBarFiOKtW/yN
AfMfoYSKuXHHJFiMx7IZdMFiN1224OnHgxtt7RCtW+NVsb2ZKZjqeIbYceN4Ex4Z
Q2JZT06+X2wO7oChrhMOaV9QujS6A0aEfCeWBOed0NNZnRVh4Xeb4rD5/yNFLdKO
kVchsNjNKDdX166g6k6fH82o8MRaTV84pd2t7Yw80ZZL0hqBt5X9AS7HKVrXQfs6
KI9WqBmTPWAQkJ9pIu5LbcHM26ayQ7MJTx/kh1pVRA97I2uid/5Q7yZLCPxblP6h
wP0hJgJEIGXaU9th0PcSErTJd27JOPkhSkRtU/EONWhJxt0E2foldrNdEYZ+USK2
a6OoBValo+ieXDFVgH/8eBm5eQoEhrmfhbOMakEbkULzD8fRIx6SIOlqeT6yWojj
8L5S6HudaTQsEq+h6w1iv6VyWrsn6VrZSNASNRGPdcsxSEsPc8CoofisO10rXULK
YmWgVqV6
=QnMV
-----END PGP SIGNATURE-----
```


</details>

MarcoFalke commented on src/test/fuzz/util.cpp:
```diff
@@ -238,10 +238,6 @@ void FillNode(FuzzedDataProvider& fuzzed_data_provider, ConnmanTestMsg& connman,
     assert(node.nServices == remote_services);
-    if (node.m_tx_relay != nullptr) {
-        LOCK(node.m_tx_relay->cs_filter);
-        assert(node.m_tx_relay->fRelayTxes == filter_txs);
```
nit in b6eae072553bd2811feed0149644a9a906489f35:

Instead of removing, I think you can do assert(peerman.getpeer(id).frelay==filter)

jnewbery replied:
`PeerManager` doesn't have a public `GetPeerRef()` method. The `Peer` objects are intentionally private to `PeerManagerImpl`.

MarcoFalke replied:
The following diff compiles for me:

```diff
diff --git a/src/test/fuzz/util.cpp b/src/test/fuzz/util.cpp
index d57c0081db..6766fbf2d9 100644
--- a/src/test/fuzz/util.cpp
+++ b/src/test/fuzz/util.cpp
@@ -225,7 +225,7 @@ void FillNode(FuzzedDataProvider& fuzzed_data_provider, ConnmanTestMsg& connman,
     const ServiceFlags remote_services = ConsumeWeakEnum(fuzzed_data_provider, ALL_SERVICE_FLAGS);
     const NetPermissionFlags permission_flags = ConsumeWeakEnum(fuzzed_data_provider, ALL_NET_PERMISSION_FLAGS);
     const int32_t version = fuzzed_data_provider.ConsumeIntegralInRange<int32_t>(MIN_PEER_PROTO_VERSION, std::numeric_limits<int32_t>::max());
-    const bool filter_txs = fuzzed_data_provider.ConsumeBool();
+    const bool relay_txs{fuzzed_data_provider.ConsumeBool()};
 
     const CNetMsgMaker mm{0};
 
@@ -241,7 +241,7 @@ void FillNode(FuzzedDataProvider& fuzzed_data_provider, ConnmanTestMsg& connman,
                 uint64_t{1},                                    // dummy nonce
                 std::string{},                                  // dummy subver
                 int32_t{},                                      // dummy starting_height
-                filter_txs),
+                relay_txs),
     };
 
     (void)connman.ReceiveMsgFrom(node, msg_version);
@@ -255,6 +255,9 @@ void FillNode(FuzzedDataProvider& fuzzed_data_provider, ConnmanTestMsg& connman,
     assert(node.nVersion == version);
     assert(node.GetCommonVersion() == std::min(version, PROTOCOL_VERSION));
     assert(node.nServices == remote_services);
+    CNodeStateStats statestats;
+    assert(peerman.GetNodeStateStats(node.GetId(), statestats));
+    assert(statestats.m_relay_txs == (relay_txs && !node.IsBlockOnlyConn()));
     node.m_permissionFlags = permission_flags;
     if (successfully_connected) {
         CSerializedNetMsg msg_verack{mm.Make(NetMsgType::VERACK)};
```

It does two things:

* Rename the `filter_txs` symbol to a better name
* Restore the check that `m_relay_txs` is correctly set after the "version msg roundtrip".

jnewbery replied:
Thanks. Done in #24692.

jnewbery commented:
Rebased

dergoegge commented on src/net_processing.cpp:
```diff
@@ -394,7 +425,7 @@ class PeerManagerImpl final : public PeerManager
 
     /** Send a version message to a peer */
-    void PushNodeVersion(CNode& pnode);
+    void PushNodeVersion(CNode& pnode, Peer& peer);
```
nit:
```suggestion
    void PushNodeVersion(CNode& pnode, const Peer& peer);
```
Or like cory suggested this could also just be a bool.

dergoegge reviewed (commented):
ACK 1066d10f71e6800c78012d789ff6ae19df0243fe - This is a good layer separation improvement with no behavior changes.

Are you still planning to address [Cory's comments](https://github.com/bitcoin/bitcoin/pull/21160#pullrequestreview-789732433)?

glozow reviewed (commented):
utACK 1066d10f71e6800c78012d789ff6ae19df0243fe

jnewbery commented:
There are still a few outstanding review comments that could be responded to or addressed in a follow-up. Leaving this here mostly as a reminder to myself:

- [ ] https://github.com/bitcoin/bitcoin/pull/21160#discussion_r736785662 "Same comment about lock scope here (and the hunk below as well)"
- [ ] https://github.com/bitcoin/bitcoin/pull/21160#discussion_r736788800 "This is the only place where these values are read outside of a lock which keeps their writes in sync. I suppose these rarely being out of sync is of no consequence?"
- [ ] https://github.com/bitcoin/bitcoin/pull/21160#discussion_r736822577 "Why not const Peer& peer here (and for MaybeSendFeefilter), since it's used read-only?"
- [ ] https://github.com/bitcoin/bitcoin/pull/21160#discussion_r736823554 "Why not make this a member function of Peer to avoid having to peek at the internal lock?"
- [ ] https://github.com/bitcoin/bitcoin/pull/21160#discussion_r736827240 "Afaics, all this needs is a bool for peer.m_tx_relay != nullptr, otherwise Peer is unused here. Maybe just pass the bool instead of requiring all of Peer?"
- [ ] https://github.com/bitcoin/bitcoin/pull/21160#discussion_r736828470 "Likewise, this doesn't need all of Peer, only a Peer::TxRelay."
- [ ] https://github.com/bitcoin/bitcoin/pull/21160#discussion_r736834651 "See comment above about changing the PushNodeVersion params, which would eliminate the need for the copies here."
- [ ] https://github.com/bitcoin/bitcoin/pull/21160#discussion_r773854573 "Instead of removing, I think you can do assert(peerman.getpeer(id).frelay==filter)"

jnewbery commented on src/net.cpp:
```diff
@@ -1044,18 +1044,11 @@ bool CConnman::AttemptToEvictConnection()
                                                HasAllDesirableServiceFlags(node->nServices),
-                                               peer_relay_txes, peer_filter_not_null, node->nKeyedNetGroup,
-                                               node->m_prefer_evict, node->addr.IsLocal(),
+                                               node->m_relays_txs.load(), node->m_bloom_filter_loaded.load(),
```
Yes, I believe this is fine. The only time both of these values are read is in ` CompareNodeTXTime()` as a tie-break between nodes that have the same `m_last_tx_time`. In the (extremely rare) window condition where a `filterclear` or `filterload` message has caused these to be written, but there's a read between `m_bloom_filter_loaded` and `m_relays_txs` being read, then the only difference is that two nodes that have the same `m_last_tx_time` might be ordered differently.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -803,6 +834,14 @@ static void PushAddress(Peer& peer, const CAddress& addr, FastRandomContext& ins
     }
 }
 
+static void AddKnownTx(Peer& peer, const uint256& hash)
```
I've been trying to keep `Peer` as a pure data structure and have all the program logic contained in the `PeerManagerImpl` functions. We've previously had logic split between net_processing functions, CNodeState methods and CNode methods, which has made it difficult to follow the program logic.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -3910,7 +3911,9 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         {
             LOCK(pfrom.m_tx_relay->cs_filter);
             pfrom.m_tx_relay->pfilter.reset(new CBloomFilter(filter));
+            pfrom.m_bloom_filter_loaded = true;
```
Done in #24692

jnewbery commented on src/net_processing.cpp:
```diff
@@ -385,7 +416,7 @@ class PeerManagerImpl final : public PeerManager
 
     /** Send a version message to a peer */
-    void PushNodeVersion(CNode& pnode, int64_t nTime);
+    void PushNodeVersion(CNode& pnode, Peer& peer, int64_t nTime);
```
Done in #24692

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1086,7 +1125,7 @@ void PeerManagerImpl::FindNextBlocksToDownload(NodeId nodeid, unsigned int count
 } // namespace
 
-void PeerManagerImpl::PushNodeVersion(CNode& pnode, int64_t nTime)
+void PeerManagerImpl::PushNodeVersion(CNode& pnode, Peer& peer, int64_t nTime)
```
Done in #24692

jnewbery commented on src/net_processing.cpp:
```diff
@@ -4420,12 +4460,12 @@ void PeerManagerImpl::MaybeSendAddr(CNode& node, Peer& peer, std::chrono::micros
 }
 
-void PeerManagerImpl::MaybeSendFeefilter(CNode& pto, std::chrono::microseconds current_time)
+void PeerManagerImpl::MaybeSendFeefilter(CNode& pto, Peer& peer, std::chrono::microseconds current_time)
```
Done in #24692

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1157,13 +1196,13 @@ void PeerManagerImpl::InitializeNode(CNode *pnode)
     }
     if (!pnode->IsInboundConn()) {
-        PushNodeVersion(*pnode, GetTime());
+        PushNodeVersion(*pnode, *peer, GetTime());
```
Done in #24692

jnewbery commented:
> There are still a few outstanding review comments that could be responded to or addressed in a follow-up. Leaving this here mostly as a reminder to myself:

I believe I've now addressed all outstanding review comments on this PR, either here or in #24692.

MarcoFalke commented:
See also #24691

jonatack commented on src/net_processing.cpp:
```diff
@@ -1330,6 +1365,14 @@ bool PeerManagerImpl::GetNodeStateStats(NodeId nodeid, CNodeStateStats& stats) c
+    } else {
+        stats.m_relay_txs = false;
+        stats.m_fee_filter_received = 0;
+    }
```
Moving this logic from CNodeStats to CNodeStateStats broke -netinfo (and perhaps other clients that expected this field to be non-null). Proposing a fix.

jonatack replied:
> Moving fRelayTxes/m_relay_txs from CNodeStats to CNodeStateStats broke -netinfo (and perhaps other clients that expected this field to be non-null). Proposing a fix.

Fixed in #25176.
